### ISSUE TITLE
Excluded file interferes with templated file

### ISSUE THREAD

**Issue by AntoineD** [copier-org/copier#1238]
### Describe the problem

If the final name of a templated file if the same as an excluded file then the template is not created. 

### Template

```shell
> cat copier.yml
_exclude:
  - /foo
  - /.git
  - /copier.yml
> ls
copier.yml  foo  foo.jinja
```

### To Reproduce

```shell
> copier copy <path to template> bar 
> ls bar

```

### Expected behavior

I would expect the templated file `foo` to be created from `foo.jinja`.

### Operating system

Linux

### Operating system distribution and version

Fedora 38

### Copier version

8.0.0

### Python version

3.11

### Installation method

pipx+pypi

**Comment by yajo** [comment]
This is the expected behavior. If you exclude foo, then it is excluded 😁

Don't exclude it if you want it included.

**Comment by pawamoy** [comment]
@yajo it wasn't super clear, but actually I think what @AntoineD describes is indeed a bug, or at least a limitation that should be documented: if you have both `foo` and `foo.jinja` files, and exclude `foo`, then `foo.jinja` is not rendered into the project.

Maybe a more concrete example would be: my template does not use a subfolder, so I have both a `README.md` file that is the readme of the template, and a `README.md.jinja` file that is the readme of the generated project. I don't want to include my template's readme, but I do want the templated readme to render.

I'm actually not sure what the behavior is when you have such files and don't exclude any. Which ones takes precedence?

**Comment by AntoineD** [comment]
@pawamoy Thank you, my use case applies to `tox.ini` files that are used to test both the template and the rendered project.

**Comment by sisp** [comment]
@AntoineD Independent of the question regarding `_exclude`, I'd always use `_subdirectory` and separate the template itself from the template project. This would avoid your problem.

**Comment by AntoineD** [comment]
> @AntoineD Independent of the question regarding `_exclude`, I'd always use `_subdirectory` and separate the template itself from the template project. This would avoid your problem.

@sisp [Unfortunately this does not work for my use case.](https://github.com/copier-org/copier/issues/1182)

**Comment by pawamoy** [comment]
Well, whether you use a subdirectory or not, you cannot enforce the name of the generated directory, so I don't think this point stands. In any case, the behavior you witnessed should be clarified in the docs, if not fixed. @sisp do you know what happens when there's both a `foo` and `foo.jinja` files? Does the second one overwrite the first, depending on the iteration order on the directory listing?

**Comment by sisp** [comment]
Ah, sorry, I hadn't read the original post carefully enough. :facepalm:

> @sisp do you know what happens when there's both a `foo` and `foo.jinja` files? Does the second one overwrite the first, depending on the iteration order of the directory listing?

I think this is the relevant section in the code:

https://github.com/copier-org/copier/blob/3a76e6645ab5d3a1c36684f8ceefd77c4fd4c6a5/copier/main.py#L639-L645

There, [`Path.iterdir()`](https://docs.python.org/3/library/pathlib.html#pathlib.Path.iterdir) determines the order. And since the docs on this method say

> The children are yielded in arbitrary order, and the special entries '.' and '..' are not included.

I'd say it's arbitrary whether the content of `foo` or the rendered content of `foo.jinja` will end up in the rendered project.

**Comment by pawamoy** [comment]
Yep, that's probably locale dependent. Thanks! Then Copier should either prevent this scenario by raising an error, or support it. Or we can document the limitation and recommend using subdirectories for such use-cases.

**Comment by yajo** [comment]
Exclusions always apply to the final rendered file name. If a file exists both with and without the .jinja suffix, the one without it is ignored.

**Comment by AntoineD** [comment]
This issue is closed but I feel like the discussion is not over and this use case may be legit as well.

What is the status?

**Comment by pawamoy** [comment]
My personal take: we need a note in the docs telling about this scenario (README.md/README.md.jinja, or tox.ini/tox.ini.jinja), that both files will be copied/rendered in an arbitrary or locale-dependent order (so not reliable), that in that case users should use a `_subdirectory` instead, and that excluding the non-templated file will not work as exclusion always apply to the final rendered file name, as pointed out by @yajo.

**Comment by yajo** [comment]
FWIW here this use case is handled: https://github.com/copier-org/copier/blob/3a76e6645ab5d3a1c36684f8ceefd77c4fd4c6a5/copier/main.py#L654-L660

Support for this predates the existence of the `_subdirectory` option. It is a valid and supported use case. It is an alternative to that option.

One example: in [this template](https://github.com/Tecnativa/doodba-copier-template/tree/d6950fbda093d31bc41fa8ffa1eddd8dd18728d6) there's a `tasks.py` file used to define tasks for template maintenance, but there's also a `tasks.py.jinja` file used to define tasks for subproject maintenance.

Maybe the docs about this feature are missing, but it is a supported and tested feature since long ago.

I closed the issue because there's nothing left to do here... If you want to render `foo.jinja`, then you need to remove `foo` from `_exclude`.

The exclusions always target the rendered name because you could be excluding files that already exist in the subdirectory. Same with `_skip_if_exists`. Maybe that can be documented too, to avoid confusion...
